### PR TITLE
Add wezterm support

### PR DESCRIPTION
**Pull request by p00f**
[wezterm](https://wezfurlong.org/wezterm/) is a GPU-accelerated, highly configurable terminal emulator.

I need help with creating the temporary file, it doesn't exist/is deleted before wezterm starts. When `load_term_conf` is set to `false`, I get:
image

`cargo test` passes and the wezterm backend works properly (other than the above issue)

### CODE REVIEW

File: src/backend/wezterm.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    }))
+}
+
+impl Wezterm {
+    fn init_args(&mut self, config: &Config) {
+        let mut fn_arg = String::from("");
+        if !&config.fonts.is_empty() {
+            self.args.push("--config".to_string());
+            fn_arg = String::from("font=require('wezterm').font_with_fallback({");
+            for font in &config.fonts {
```
Comment (p00f) [inline src/backend/wezterm.rs:31]: #40 I'm new to rust and I feel this might not be the best way to do it.

File: src/backend/wezterm.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                fn_arg.push('\"');
+                fn_arg.push_str(font);
+                fn_arg.push_str("\",");
```
Comment (p00f) [inline src/backend/wezterm.rs:34]: #40 I'm new to rust and I feel this might not be the best way to do it.

File: src/backend/wezterm.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+impl Wezterm {
+    fn init_args(&mut self, config: &Config) {
+        let mut fn_arg = String::from("");
+        if !&config.fonts.is_empty() {
+            self.args.push("--config".to_string());
+            fn_arg = String::from("font=require('wezterm').font_with_fallback({");
+            for font in &config.fonts {
+                fn_arg = fn_arg + "\"" + &font[..] + "\",";
```
Comment (p00f) [inline src/backend/wezterm.rs:32]: #40 I'm new to rust and I feel this might not be the best way to do it.

File: src/backend/wezterm.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            }
+            fn_arg.push_str("})");
```
Comment (p00f) [inline src/backend/wezterm.rs:34]: #40 I'm new to rust and I feel this might not be the best way to do it.

**Comment by p00f** [comment]
@beeender I'm converting this PR to a draft, pinging you so that you are notified

**Comment by beeender** [comment]
Nice! Thanks for adding this! Just let me know when it is done.

**Comment by p00f** [comment]
@beeender I need some help here, it seems the temporary file is not created even though I mostly copied that from `src/backend/kitty.rs`. The file is a blank config (ie `return {}` in Lua) which is created and used only when `load_term_conf` is false

**Comment by beeender** [comment]
I did some testing with your branch and the temp file worked.

What is your system? I am using archlinux.

```
➜  glrnvim git:(p00f/master) RUST_LOG=debug cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/glrnvim`
[2021-05-14T02:26:00Z DEBUG glrnvim] Start command: "/usr/bin/wezterm" "--config" "font=require(\'wezterm\').font_with_fallback({ \"FiraCode Nerd Font\",})" "--config" "font_size=12" "--config-file" "/tmp/.tmp4HcJVy" "start" "--class" "glrnvim" "--" "nvim" "+set termguicolors" "+set title" "--cmd" "let g:glrnvim_gui=1"
➜  glrnvim git:(p00f/master)  2021-05-14T02:26:00.497Z INFO  wezterm_mux_server_impl::local > setting up /run/user/1000/wezterm/gui-sock-346845
 2021-05-14T02:26:00.573Z INFO  wezterm_gui::termwindow        > OpenGL initialized! Radeon RX 580 Series (POLARIS10, DRM 3.40.0, 5.12.1-zen2-1-zen, LLVM 11.1.0) OpenGL ES 3.2 Mesa 21.0.3 is_context_loss_possible=false wezterm version: 20210502-154244-3f7122cb
 2021-05-14T02:26:00.595Z WARN  mux::localpane                 > unknown DeviceControlMode::Enter EnterDeviceControlMode(params: [], intermediates: [], byte: 't' 0x74, ignored_extra_intermediates=false)
 2021-05-14T02:26:00.595Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 6d m
 2021-05-14T02:26:00.595Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 75 u
 2021-05-14T02:26:00.595Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 78 x
 2021-05-14T02:26:00.595Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 3b ;
➜  glrnvim git:(p00f/master)  2021-05-14T02:26:01.045Z WARN  wezterm_font                   > No fonts contain glyphs for these codepoints: \u{2105}\u{2630}\u{33d1}.
Placeholder 'Last Resort' glyphs are being displayed instead.
You may wish to install additional fonts, or adjust your
configuration so that it can find them.
https://wezfurlong.org/wezterm/config/fonts.html has more information about configuring fonts.
Set warn_about_missing_glyphs=false to suppress this message.

➜  glrnvim git:(p00f/master) cat /tmp/.tmp4HcJVy
return {}
➜  glrnvim git:(p00f/master)
```

**Comment by p00f** [comment]
I'm using archlinux too

**Comment by p00f** [comment]
It works when I do `cargo run` but not when I do `cargo build` then `./target/debug/glrnvim`

**Comment by p00f** [comment]
I cloned it to a fresh directory, `cargo build` then `./target/debug/glrnvim` works now

**Comment by beeender** [comment]
No idea ... maybe something related cargo cache?

**Comment by p00f** [comment]
> No idea ... maybe something related cargo cache?

Yeah `cargo clean` fixed it

**Comment by beeender** [comment]
BTW, I noticed when start glrnvim with wezterm, it automatically attach to the current tmux/screen session. That is not the desired behavior. Is there an option for wezterm to prevent that?

**Comment by p00f** [comment]
I don't use tmux so I don't understand, is this what you mean?

https://user-images.githubusercontent.com/36493671/118216387-07db4400-b491-11eb-9c69-c27d8eeb3b37.mp4

**Comment by beeender** [comment]
image

Uh, no,, it is not the current tmux/screen session. I guess it is the built-in multiplexer. It looks strange to have that enabled with a VIM gui. Can that be disabled?

**Comment by p00f** [comment]
Oh sure one minute

**Comment by p00f** [comment]
Fixed

**Comment by beeender** [comment]
I got those warning messages
```
➜  glrnvim git:(p00f/master)  2021-05-14T03:25:01.678Z INFO  wezterm_mux_server_impl::local > setting up /run/user/1000/wezterm/gui-sock-501628
 2021-05-14T03:25:01.758Z INFO  wezterm_gui::termwindow        > OpenGL initialized! Radeon RX 580 Series (POLARIS10, DRM 3.40.0, 5.12.1-zen2-1-zen, LLVM 11.1.0) OpenGL ES 3.2 Mesa 21.0.3 is_context_loss_possible=false wezterm version: 20210502-154244-3f7122cb
 2021-05-14T03:25:01.777Z WARN  mux::localpane                 > unknown DeviceControlMode::Enter EnterDeviceControlMode(params: [], intermediates: [], byte: 't' 0x74, ignored_extra_intermediates=false)
 2021-05-14T03:25:01.777Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 6d m
 2021-05-14T03:25:01.777Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 75 u
 2021-05-14T03:25:01.777Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 78 x
 2021-05-14T03:25:01.777Z WARN  mux::localpane                 > unhandled DeviceControlMode::Data 3b ;
➜  glrnvim git:(p00f/master)  2021-05-14T03:25:02.223Z WARN  wezterm_font                   > No fonts contain glyphs for these codepoints: \u{2105}\u{2630}\u{33d1}.
Placeholder 'Last Resort' glyphs are being displayed instead.
You may wish to install additional fonts, or adjust your
configuration so that it can find them.
https://wezfurlong.org/wezterm/config/fonts.html has more information about configuring fonts.
Set warn_about_missing_glyphs=false to suppress this message.
```
My config
```
# vim:fileencoding=utf-8:ft=yaml

# Choose the backend terminal to run neovim in.
# Current supported terminals: alacritty, urxvt, kitty.
#
#backend: alacritty
backend: wezterm


# The fonts to be used. Multi fonts can be supplied.
# The first one will be set as the major font. Others will be set as
# fallback fonts according to the given orders if possible.
# NOTE: Not all backends support fallback font.
fonts:
  #    - FiraCode Nerd Font
    - JetBrains Mono
#   - Mononoki
#   - monofur

# The font size to be used.
#
#font_size: 12
load_term_conf: false
```

Are those warning messages expected?

**Comment by p00f** [comment]
Yeah, the config you supplied doesn't have the glyphs need to display what is on your screen

**Comment by p00f** [comment]
If you have desktop notifications you get a notification about this

File: src/backend/wezterm.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        self.temp_file = Some(file);
+    }
+}
+
+impl Functions for Wezterm {
+    fn create_command(&mut self, config: &Config) -> std::process::Command {
+        self.init_args(config);
+        let mut command = std::process::Command::new(self.exe_path.to_owned());
+        command.args(&self.args);
+        if !config.load_term_conf {
```
Comment (beeender) [inline src/backend/wezterm.rs:61]: Backend `alacritty` and `kitty` support "merge" the config with the terminal's config.
So if user specifies a font in the glrnvim's config and also has `load_term_conf = true`, the glrnvim will load the default terminal's config and merge it with what user set in the glrnvim's config.

Can this be easily supported by wezterm backend?

Comment (p00f) [reply src/backend/wezterm.rs:61]: Yeah, in this case glrnvim's config will simply override the font and rest of the config will be used from the user's wezterm config

https://wezfurlong.org/wezterm/config/files.html#configuration-overrides

**Review by beeender — APPROVED** [review]

**Comment by beeender** [comment]
Thanks for adding this! Merged!

**Comment by beeender** [comment]
@all-contributors please add @p00f for code

**Comment by p00f** [comment]
Can you please cut a release? I've written my own PKGBUILD for now

**Comment by beeender** [comment]
sure!

**Comment by beeender** [comment]
@p00f `1.2.0` has been released. And the glrnvim in aur has been updated as well.
https://aur.archlinux.org/packages/glrnvim/

**Comment by p00f** [comment]
Thanks!

**Pull request merged**